src/concurrent_cache.h

```diff
@@ -30,6 +30,31 @@
 namespace zim
 {
 
+template<typename CostEstimation>
+struct FutureToValueCostEstimation {
+  template<typename T>
+  static size_t cost(const std::shared_future<T>& future) {
+    // The future is the value in the cache.
+    // When calling getOrPut, if the key is not in the cache,
+    // we add a future and then we compute the value and set the future.
+    // But lrucache call us when we add the future, meaning before we have
+    // computed the value. If we wait here (or use future.get), we will dead lock
+    // as we need to exit before setting the value.
+    // So in this case, we return 0. `ConcurrentCache::getOrPut` will correctly increase
+    // the current cache size when it have an actual value.
+    // We still need to compute the size of the value if the future has a value as it
+    // is also use to decrease the cache size when the value is drop.
+    using namespace std::chrono_literals;
+    std::future_status status = future.wait_for(0s);
+    if (status == std::future_status::ready) {
+      return CostEstimation::cost(future.get());
+    } else {
+      return 0;
+    }
+  }
+
+};
```

Comment:
I think that such a tweak is not needed. Even if it addresses some scenarios, we better avoid it. You can use a dummy cost estimation config that returns 0 unconditionally. The change in `ConcurrentCache::getOrPut()` should be enough.

Reply:
The `cost` method is also called when we remove the value from the cache. If we use a dummy cost of 0 all the time, `ConcurrentCache` would have to decrease the cache size itself.

Reply:
As far as I understand the current assumption is that the cost of a cached item doesn't change, otherwise the total cost maintained by the cache will be wrong (in particular it may underflow and make the cache stop functioning). Then the cost value returned during addition to the cache can be recorded with the item. This will protect against the said potential problem if the cost of an item happens to depend on its internal state, at the much more acceptable cost (pun intended) of the cache not correctly respecting the cost limit. Another effect will be a little gain in performance if cost computation is costly.